src/Build/Evaluation/ProjectRootElementCacheBase.cs

```diff
@@ -0,0 +1,126 @@
+
+        /// <summary>
+        /// Delegate for StrongCacheEntryRemoved event
+        /// </summary>
+        internal delegate void StrongCacheEntryRemovedDelegate(object sender, ProjectRootElement projectRootElement);
+
+        /// <summary>
+        /// Callback to create a ProjectRootElement if need be
+        /// </summary>
+        internal delegate ProjectRootElement OpenProjectRootElement(string path, ProjectRootElementCacheBase cache);
+
+        /// <summary>
+        /// Event that is fired when an entry in the Strong Cache is removed.
+        /// </summary>
+        internal static event StrongCacheEntryRemovedDelegate StrongCacheEntryRemoved;
+
+        /// <summary>
+        /// Event which is fired when a project root element is added to this cache.
+        /// </summary>
+        internal event ProjectRootElementCacheAddEntryHandler ProjectRootElementAddedHandler;
+
+        /// <summary>
+        /// Event which is fired when a project root element in this cache is dirtied.
+        /// </summary>
+        internal event EventHandler<ProjectXmlChangedEventArgs> ProjectRootElementDirtied;
+
+        /// <summary>
+        /// Event which is fired when a project is marked dirty.
+        /// </summary>
+        internal event EventHandler<ProjectChangedEventArgs> ProjectDirtied;
+
+        internal abstract ProjectRootElement Get(string projectFile, OpenProjectRootElement openProjectRootElement,
+            bool isExplicitlyLoaded,
+            bool? preserveFormatting);
+
+        internal abstract void AddEntry(ProjectRootElement projectRootElement);
+
+        internal abstract void RenameEntry(string oldFullPath, ProjectRootElement projectRootElement);
+
+        internal abstract ProjectRootElement TryGet(string projectFile);
+
+        internal abstract ProjectRootElement TryGet(string projectFile, bool? preserveFormatting);
+
+        internal abstract void DiscardStrongReferences();
+
+        internal abstract void Clear();
+
+        internal abstract void DiscardImplicitReferences();
+
+        internal abstract void DiscardAnyWeakReference(ProjectRootElement projectRootElement);
+
+        /// <summary>
+        /// Raises the <see cref="ProjectRootElementDirtied"/> event.
+        /// </summary>
+        /// <param name="sender">The dirtied project root element.</param>
+        /// <param name="e">Details on the PRE and the nature of the change.</param>
+        internal void OnProjectRootElementDirtied(ProjectRootElement sender, ProjectXmlChangedEventArgs e)
+        {
+            var cacheDirtied = ProjectRootElementDirtied;
+            cacheDirtied?.Invoke(sender, e);
+        }
+
+        /// <summary>
+        /// Raises the <see cref="ProjectDirtied"/> event.
+        /// </summary>
+        /// <param name="sender">The dirtied project.</param>
+        /// <param name="e">Details on the Project and the change.</param>
+        internal void OnProjectDirtied(Project sender, ProjectChangedEventArgs e)
```

Comment:
Maybe override these in the simple cache and throw, to assert the simple cache does not account for project mutations?